### ISSUE TITLE
`wp_safe_redirect` does not support HTTP response codes 400 and up

### ISSUE THREAD

**Issue by helen** [tlovett1/safe-redirect-manager#241]
As reported in #239, coming out of #205. This issue is to make sure we document the base issue and work through solutions before deciding what direction to go with this plugin.

As of WordPress 5.4, `wp_safe_redirect()` and `wp_redirect()` only allow for `3xx` response codes ([commit](https://github.com/WordPress/wordpress-develop/commit/40946e984a5c6b775a0e34d86b36a004c6a70204)). This does seem correct to me, as other codes are not relevant to redirects. However, in #205 a very understandable use case for managing `4xx` codes (specifically `410`) is outlined:

<hr />
**You wish to remove a page from your website.**

* Step 1 - Set up a 301 for that URL to a new page with new info.
* Step 2 - Work to remove any links to that page. For a test page, this may take minutes. Or it may take months for a popular page (perhaps you should rebuild and republish popular URLs?)
* Step 3 - Set up a 410 Gone for that URL.
* Step 4 - The 410 Gone will appear in your Google Search Console.
* Step 5 - Remove the 410 Gone for that URL.

Second Results: When you remove the 301 there is no 404 in your Google Search Console.
<hr />

What should Safe Redirect Manager do here to support its most common use cases while still remaining, at heart, a redirect manager and not an HTTP code manager?

**Comment by helen** [comment]
Commenting with my personal thoughts:

At first, I thought we just shouldn't support non-`3xx` codes at all, but I think I understand the use case better now and it does seem like a good idea to support specific codes to help with managing the outcome of changing redirects. That said, I don't think we should actually allow for redirects on the `4xx` series as proposed in #239, because they are not redirects. I am not yet strongly opinionated about what it should do in terms of displaying a given template, so that's where I would point the discussion - should SRM show a `wp_die()`-style message with the appropriate response code and leave displaying other templates to a hook, or should it support at least `404.php`?

**Comment by dinhtungdu** [comment]
Done some research and I think we should support only `3xx` code out of the box (in the other words, remove current 4xx codes).

SRM has `srm_do_redirect` action which developers can use to do custom redirects if needed. Changes in #239 can make use of that action for example.

https://github.com/10up/safe-redirect-manager/blob/b0e8e0ed818f583a7819805bc333c563070a4bd9/inc/classes/class-srm-redirect.php#L229-L234

cc @aosmichenko @PopVeKind

**Comment by helen** [comment]
FWIW I do see the use case for supporting certain `4xx` codes out of the box, making a decision here would definitely benefit from more insight into typical use cases for those who do use that functionality either in SRM or another redirection plugin.

**Comment by dustinrue** [comment]
410 feels like a natural extension of what SRM is tasked with doing which is to help manage the lifecycle of content. Sometimes mistakes are made and a redirect is used to correct a URL. Sometimes content is moved due to $reason. Sometimes content just needs to be gone. 

It may be worth considering, however, how other search engines treat these status codes.

**Comment by ciprianimike** [comment]
from the point of view of search engines a `404 Not found` and a `410 Gone` have no differences, except that URLs that return 410 might be de-indexed faster.

So the second step in the description, 404 should take just some more time to be reported in the Google Search Console.

> Second Results: When you remove the 301 there is no 404 in your Google Search Console.

I would agree with @dinhtungdu, SRM should remain focused solely on redirects, therefore on 3xx statuses, without providing workarounds out of the box, which can be implemented with the `srm_do_redirect` action.

**Comment by dkotter** [comment]
**tldr;** I propose we add proper support for 403, 404 and 410 error codes, since we already allow those to be selected when creating redirects. Full thoughts below:

Not to add even more voices here but spent some time looking this over and putting some thought into this.

From what I can tell, we have for some time (if not since the beginning) "supported" the 403 and 404 status codes when adding redirects and we recently added "support" for 410 codes. As @helen mentions though, the underlying `wp_redirect` method had support removed for any codes above 399 in the WP 5.4 release. Since that release, if you set up a redirect with a 403, 404 or 410 code, all that happens is the error message `HTTP redirect status code must be a redirection code, 3xx.` is shown (and the proper status code is not used).

This means those status types don't actually do what people would expect them to do, assuming they are running WP 5.4+. I agree then with both @dinhtungdu and @ciprianimike that if those aren't working, at the very least we should remove those. I can also understand the thought that this feels like maybe going outside the scope of this plugin to support those, since we wouldn't redirect if those status codes are used.

All that said, because we have technically "supported" those codes for a while (and this means some people may be using those even though they don't work so backwards compat to think about) my vote would be to add actual support for those, so the functionality we advertise actually works (instead of just removing that functionality).

This leads back to @helen's question around what should happen if a 403, 404 or 410 status code is used. Makes sense to me to show the default `404.php` template if 404 is used and then probably just use a `wp_die` error message for the other two (assuming we can do that and still output the proper status code). I could easily be persuaded to use the default `404.php` template for all of those though. I also like the idea of ensuring we have a hook in place to easily allow template overrides here.

Also, since we don't want to redirect on those codes, I think we could implement some admin UI here to hide the _redirect_to_ field if a 4** code is used. Could be as simple as they select 403, 404 or 410 and we then hide the field. Could also maybe rearrange the UI a bit more to where you select a code first and then get the appropriate fields, but not sure it's worth that effort. Just seems weird to allow a _redirect_to_ URL if we aren't actually redirecting anything.

**Comment by jeffpaul** [comment]
Making note that the plugin currently requires at least WP version 5.7, so anyone running a current version of SRM is already seeing the changes introduced in WP 5.4.

https://github.com/10up/safe-redirect-manager/blob/a7f855f44187927648eb9b0e02cd13b7346fce3d/readme.txt#L4

I'm in full agreement with @dkotter's assessment here, with a tweak on the UI where if someone selects a 4xx code that we do not hide the `redirect_to` but instead disable that input field and instead show a message to the effect of "No `redirect_to` value applies for 4xx error codes.".
